File: src/CompileTime/tokenizer.cpp
#include "Tokenizer.hpp"
#include <iostream>
#include <string>
#include <vector>
using namespace std;
Tokenizer::Tokenizer(string code)
  :code(code),
  line(1),
  colmun(0)
{}

Token::Token(string value, TokenType type, int line, int colmun) 
  :value(value), 
  type(type),
  line(line),
  colmun(colmun)
{}


char Tokenizer::take() {
  if(code.size() <= 0) {
    return ' ';
  }
  char prev = code[0];
  code.erase(0, 1);
  return prev;
}

char Tokenizer::at() {
  if(code.size() <= 0) {
    return ' ';
  }
  return code[0];
}

bool Tokenizer::isNum() {
  string nums = "0123456789";
  // find() returns npos if not found
  return nums.find(this->at()) != string::npos;
}

Token Tokenizer::set(string value, TokenType type) {
  Token tok = Token(value, type, line, colmun);

  this->current_token = tok;
  return tok;
}

Token Tokenizer::tokenize() {
  //take skippable chars && throws them
  while(this->at() == ' ' || this->at() == '\t') this->take();

  if(code.size() <= 0) return set("END", TokenType::eof);
  switch(this->at()) {
      default:
        cout << "invaild char" << endl;
        break;
      case '0':
      case '1':
      case '2':
      case '3':
      case '4':
      case '5':
      case '6':
      case '7':
      case '8':
      case '9':
        string res;
        while(isNum()) {
          res += this->take();
        }
        return this->set(res, TokenType::number);
    }
  return set("END", TokenType::eof); 
}

File: src/CompileTime/tokenizer.fs
namespace AtomicLang
module lexer =

  type TokenType =
  | Num
  | Null
  | EOF
  [<StructuredFormatDisplay("line: {line}, colmun: {colmun}, type: {ttype}, value: {value}")>]
  type Token(value : string, ttype : TokenType, line : int, colmun : int) =
    member this.value = value;
    member this.ttype = ttype;
    member this.line = line;
    member this.colmun = colmun;

  type Tokenizer(code : string) as self = 
    let mutable line = 1;
    let mutable colmun = 0;

    let mutable code = code;
    
    let take() : char =
      if code.Length <= 0 then
        ' ';
      else
        colmun <- colmun + 1;
        let prev = code[0];
        printfn "%c"prev
        code <- code.Substring(1);
        prev;

    let at() : char = 
      if code.Length <= 0 then 
        ';';
      else
        code[0];
    let isNum() : bool = 
      "0123456789".Contains code[0];

    let mutable ccurrent_token : Token = new Token("unknown", TokenType.Null, -1, -1)
    member this.current_token() = ccurrent_token;

    member this.getLine with get () = line
    member this.getColmun with get () = colmun

    member x.set(value : string, ttype : TokenType) = 
      ccurrent_token <- new Token(value, ttype, line, colmun); 
      printfn "%A" (x.current_token())
      ignore

    member this.tokenize() : Token =
      while code.Length > 0 && at() = ' ' do take() |> ignore
      while code.Length > 0 && at() = '\n' do
        take() |> ignore
        line <- line + 1
        colmun <- 0
      if code.Length <= 0 then this.set("END", TokenType.EOF) |> ignore
      else
        match at() with
        | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '0' ->
          let mutable res = "";
          while code.Length > 0 && isNum() do
            res <- res + string(take())
            printfn "%s"res
          this.set(res, TokenType.Num) |> ignore
        | _ ->
          if at() = ' ' || at() = '\n' || at() = '\t' then
            this.tokenize() |> ignore
          else
            printfn "unexcepted char %c." (take());
      this.current_token();
